### ISSUE TITLE
A bug of integrate function

### ISSUE THREAD

**Issue by MakotoTsukada** [sympy/sympy#22033]
When we normalize the 2nd order Chebyshev polynomial x^2 - 1/4 of the 2nd kind by the L^2 norm on [-1, 1] with density function sqrt(1 - x^2)dt, we encounter a wrong result.
```
>>> from sympy import *
>>> from sympy import __version__
>>> __version__
'1.8'
>>> x = Symbol('x')
>>> f = (x**2 - Rational(1, 4))**2 * sqrt(1 - x**2); f
sqrt(1 - x**2)*(x**2 - 1/4)**2
>>> integrate(f, (x, -1, 1))
0
```
The norm is 0?.
The integrand of the above definite integral is strictly positive, so it is a bug.

If we divide the interval [-0, 1] into [-1, 0] and [0, 1], we have correct results as follows:
```
>>> integrate(f, (x, -1, 0))
pi/64
>>> integrate(f, (x, 0, 1))
pi/64
```
And, expanding the integrand, we also have the correct.
```
>>> g = f.expand(); g
x**4*sqrt(1 - x**2) - x**2*sqrt(1 - x**2)/2 + sqrt(1 - x**2)/16
>>> integrate(g, (x, -1, 1))
pi/32
```
This bug exists since version 1.5 of SymPy.
```
>>> from sympy import *
>>> from sympy import __version__
>>> __version__
'1.4'
>>> x = Symbol('x')
>>> f = (x**2 - Rational(1, 4))**2 * sqrt(1 - x**2); f
sqrt(1 - x**2)*(x**2 - 1/4)**2
>>> integrate(f, (x, -1, 1))
pi/32
```
This is correct.

**Comment by jksuom** [comment]
The result of the manual integrator may depend on the assumptions on the integration variable:
```
>>> x = Symbol('x')
>>> f = (x**2 - Rational(1, 4))**2 * sqrt(1 - x**2)
>>> piecewise_fold(integrate(f, x))
Piecewise((-x**3*(1 - x**2)**(3/2)/6 + x*sqrt(1 - x**2)/32 + asin(x)/32, (x > -1) & (x < 1)))
>>> xr = Symbol('xr', real=True)
>>> fr = (xr**2 - Rational(1, 4))**2 * sqrt(1 - xr**2)
>>> piecewise_fold(integrate(fr, xr))
Piecewise((-xr**2*(1 - xr**2)**(3/2)*Abs(xr)/6 + sqrt(1 - xr**2)*Abs(xr)/32 - asin(sqrt(1 - xr**2))/32, (xr > -1) & (xr < 1) & Ne(xr, 0)))
```
The latter antiderivative is symmetric, which is wrong.

In the integral over the interval [-1, 1], the real variable comes from last one  of these lines:
https://github.com/sympy/sympy/blob/6ef14a954288079cde3d1bf304189f6b8f9296dc/sympy/integrals/integrals.py#L472-L479
There was another case like this earlier: #21671.

The integrals over the intervals [-1, 0] and [0, 1], where a negative resp. positive dummy is substituted, seem to be correct.

**Comment by MakotoTsukada** [comment]
We wrote primitive functions of f and fr of the above comment.

[sympy_test2 - Jupyter Notebook.pdf](https://github.com/sympy/sympy/files/7116593/sympy_test2.-.Jupyter.Notebook.pdf)

**Comment by oscarbenjamin** [comment]
> The result of the manual integrator may depend on the assumptions on the integration variable:

I'm not sure how to debug this. Do you think that the bug is in the manual integrator or something else?

**Comment by MakotoTsukada** [comment]
I'm not sure about the inside of integrate function, but in mathematics it is easy to see that the above integration is equivalent to the following integration by substitution.
```
>>> integrate('(sin(t)**2-1/4)**2*cos(t)**2', (t, '-pi/2', 'pi/2'))
pi/32
```
This is also correct.

**Comment by oscargus** [comment]
> Do you think that the bug is in the manual integrator or something else?

My guess is that some function gets rewritten/evaluated when the argument is real/extended real and either the rewrite is wrong or the new expression cannot be integrated correctly.

There are just three cases of checking for extended_real in `manualintegrate`, in the quadratic denominator and erf rules. Not even sure that the checks are for the symbol though. So most likely it is a matter of a rewrite at some stage.

"Worst case" would be to do the opposite transformation in `manualintegrate`, replace all variables with a Dummy without any assumptions as there are cases where one would naturally put the real assumption to a variable outside of the actual integrate call, so it is not just the internal addition of real that is the problem.

Isn't there a version that actually outputs the computations step-by-step in some old SympyLive? I seem to recall checking out that code a number of years back. Having that up and running would probably make the debugging quite a lot easier.

**Comment by oscarbenjamin** [comment]
There is `integral_steps`. Also setting the `SYMPY_DEBUG` environment variable shows output from the integration routines.

**Comment by oscarbenjamin** [comment]
In #21671 the issue seems to be a substitution like `y = x**2` which is problematic if the integral is over positive and negative `x`. The effect of `extended_real` is something like this:
```python
In [13]: sqrt(xr**2)
Out[13]: │xr│

In [14]: sqrt(x**2)
Out[14]: 
   ____
  ╱  2 
╲╱  x  
```

A lot of output can be seen from this (too much to show):
```python
In [8]: from sympy.integrals.manualintegrate import manualintegrate, integral_steps

In [9]: xr = Symbol('xr', real=True)

In [10]: fr = (xr**2 - Rational(1, 4))**2 * sqrt(1 - xr**2)

In [11]: integral_steps(fr, xr)
```

**Comment by oscargus** [comment]
Ahh, that is right! What I seems to recall is some function that turns that output into HTML or LaTeX so that it is more easily readable. But it didn't work straight out of the box three(?) years ago, so probably doesn't work now either...

**Comment by oscargus** [comment]
Here is what the difference seems to be.

With real variable:
```
List of Alternative Rules
Rule 1: <function rewrites_rule at 0x7f2a31605430>
Rule 2: <function substitution_rule at 0x7f2a31603c10>
substituted: -(1 - xr**2)*(xr**2 - 1/4)**2/xr, u: sqrt(1 - xr**2), u_var: _u
...
substituted: -sqrt(1 - xr**2)*(xr**2 - 1/4)**2/(2*xr), u: 1 - xr**2, u_var: _u
substituted: sqrt(1 - xr**2)*(xr**2 - 1/4)/(4*xr), u: (xr**2 - 1/4)**2, u_var: _u
substituted: sqrt(1 - xr**2)*(xr**2 - 1/4)**2/(2*xr), u: xr**2 - 1/4, u_var: _u
substituted: sqrt(1 - xr**2)*(xr**2 - 1/4)**2/(2*xr), u: xr**2, u_var: _u
List of Substitution Rules
```

With general variable:
```
List of Alternative Rules
Rule 1: <function rewrites_rule at 0x7f15cdde7430>
Rule 2: <function substitution_rule at 0x7f15cdde6c10>
substituted: -(1 - x**2)*(x**2 - 1/4)**2/x, u: sqrt(1 - x**2), u_var: _u
...
substituted: -sqrt(1 - x**2)*(x**2 - 1/4)**2/(2*x), u: 1 - x**2, u_var: _u
substituted: sqrt(1 - x**2)*(x**2 - 1/4)/(4*x), u: (x**2 - 1/4)**2, u_var: _u
substituted: sqrt(1 - x**2)*(x**2 - 1/4)**2/(2*x), u: x**2 - 1/4, u_var: _u
substituted: sqrt(1 - x**2)*(x**2 - 1/4)**2/(2*x), u: x**2, u_var: _u
Rule 3: <function condition.<locals>.conditioned_rl at 0x7f15cdab80d0>
Integral: sqrt(1 - x**2)*(x**2 - 1/4)**2 is rewritten with <function <lambda> at 0x7f15cdde6d30> on symbol: x
Rule 4: <function condition.<locals>.conditioned_rl at 0x7f15cdab81f0>
Integral: sqrt(1 - x**2)*(x**2 - 1/4)**2 is rewritten with <function <lambda> at 0x7f15cdde6ee0> on symbol: x
```

So Rule 3 and Rule 4 are not applied.

More or less from this step, all involved variables are `Dummy`, so they should not have any assumptions.

Questions: 
1. Why are Rules 3 and 4 not triggered for the real variable?
2. Why are the results incorrect if these rules are not triggered?

**Comment by jksuom** [comment]
It seems that there is a bug in `Pow._eval_power`.
```
>>> (1/x).subs(x**2, y)
1/x
>>> xr = Symbol('xr', real=True)
>>> (1/xr).subs(xr**2, y)
1/sqrt(y)
```
These lines make all real valued base expressions nonnegative: https://github.com/sympy/sympy/blob/3794c3e801a2beb2e97a565228b0d2b88fa5afa6/sympy/core/power.py#L425-L426

**Comment by oscargus** [comment]
Good find! It's been around for seven years as far as I can tell. Will be interesting to see what the consequences are of fixing it...

**Comment by oscarbenjamin** [comment]
I think that `_eval_power` is fine:
```
(Pdb) p self
xr**2
(Pdb) p other
-1/2
(Pdb) retval
1/Abs(xr)
```
The bug is in `_eval_subs` which checks the result of `_eval_power`:
https://github.com/sympy/sympy/blob/3794c3e801a2beb2e97a565228b0d2b88fa5afa6/sympy/core/power.py#L835-L837
It accepts `1/Abs(xr)` as a Pow but I expect that is not the intention of checking for `Pow` there.

**Comment by oscargus** [comment]
I'd guess that #18527 is also a case of this.

**Comment by jksuom** [comment]
It is not true in general that `(b**e)**f == b**(e*f)`. The purpose of `_check` is to find out when this is valid (`combines is True`).

In SymPy, `b**e` is defined as `exp(e*log(b))` where `log(b)` denotes the principal value. Then the principal value `log(b**e)` is `e*log(b) + n*2*pi*I` for some integer `n`. Hence

    (b**e)**f == exp(e*f*log(b) + f*n*2*pi*I),

which is different from `b**(e*f) == exp(e*f*log(b))` unless `f*n` is an integer.

There are two cases when `f*n` is an integer:

1. `f ` is an integer.
This is handled by `as_int(pow, strict=False)` in `_check`.

2. `b` is positive and `e` is real.
Then `b**e` is positive, so `log(b**e)` and `e*log(b)` are both real, and therefore `n` is 0. 

I think that something like this could be used in the latter case:

                        combines = (terms1.is_nonnegative and coeff1.is_real or
                            isinstance(Pow._eval_power(
                            Pow(*old.as_base_exp(), evaluate=False),
                            pow), (exp, Symbol)))

**Comment by oscarbenjamin** [comment]
If the condition is that `b ** e` is positive would it not be better to check that directly like `if (b**e).is_positive`?

**Comment by jksuom** [comment]
I think that both `b` and `e` should be checked. `b**e` could be positive even if neither `b` nor `e` satisfies the conditions.

**Comment by oscarbenjamin** [comment]
Okay so `if b.is_positive and e.is_real`?

**Comment by jksuom** [comment]
We could accept `b` taking value 0. It could be a symbol with `nonnegative=True`.

**Comment by oscarbenjamin** [comment]
Then a negative exponent couldn't be allowed I guess. So `if (b.is_positive and e.is_real) or (b.is_nonnegative and e.is_nonnegative)` or perhaps:
```python
if ((b.is_extended_positive and e.is_extended_real)
    or (b.is_extended_nonnegative and e.is_extended_nonnegative)):
```

**Comment by jksuom** [comment]
Well, there will probably never appear `0**e` in practice, only something like `b**e` where `b` is nonnegative. But, to be on the safe side, we are likely not losing much if we take `b.is_positive and e.is_real`. I don't think that the extended versions would be of much use.

**Comment by jksuom** [comment]
It seems that the inner power is given by `ct2` in `_check`, not by `ct1` as I had assumed, so it looks like the conditions should be

    terms2.is_positive and coeff2.is_real

(`terms1 == terms2` so it does not matter which base to use, but the exponent should be the inner one, i.e., that of `old`.)

**Comment by MakotoTsukada** [comment]
Question: Please read the attached file.
[sympy.pdf](https://github.com/sympy/sympy/files/7128974/sympy.pdf)

**Comment by oscarbenjamin** [comment]
These are the examples that hit this particular codepath in the core tests:
```python
from sympy import *

x, y = symbols('x y')
xr, yr = symbols('x y', real=True)
xnn, ynn = symbols('x y', nonnegative=True)
xc, yc = symbols('x y', complex=True)
w = Symbol('w')
c2, c3, q1p, q2p, c1, s1, s2, s3 = symbols('c2 c3 q1p q2p c1 s1 s2 s3')

# combine=True examples:

assert ((3**(1 + x) + 2**(1 + x))/(3**x + 2**x)).subs(exp(x*log(Rational(2))), w) == (2**(x + 1) + 3**(x + 1))/(2**x + 3**x)
assert exp(I*xr).subs(exp(xr), y) == y**I
assert (2**(I*xr)).subs(2**xr, y) == y**I
eq = ((-2)**(I*xr))
assert ((-2)**(I*xr)).subs((-2)**xr, y) == ((-2)**(I*xr))
assert (xnn**4 + xnn**3 + xnn**2 + xnn + sqrt(xnn)).subs(xnn**2, ynn) == \
    ynn**Rational(1, 4) + ynn**Rational(3, 2) + sqrt(ynn) + ynn**2 + ynn
assert xnn.subs(xnn**3, ynn) == ynn**Rational(1, 3)

# combine=False examples:

test = (c2**2*q2p*c3 + c1**2*s2**2*q2p*c3 + s1**2*s2**2*q2p*c3 - c1**2*q1p*c2*s3 - s1**2*q1p*c2*s3)
assert (test.subs({c1**2: 1 - s1**2, c2**2: 1 - s2**2, c3**3: 1 - s3**2})
    == c3*q2p*(1 - s2**2) + c3*q2p*s2**2*(1 - s1**2) - c2*q1p*s3*(1 - s1**2) + c3*q2p*s1**2*s2**2 - c2*q1p*s3*s1**2)

p = 6*xr**5 + xr**4 - 4*xr**3 + 4*xr**2 - 2*xr + 3
p_subbed = 6*xr**5 - 4*xr**3 - 2*xr + yr**4 + 4*yr**2 + 3
assert p.subs([(xr**i, yr**i) for i in [2, 4]]) == p_subbed

p = 6*xc**5 + xc**4 - 4*xc**3 + 4*xc**2 - 2*xc + 3
p_subbed = 6*xc**5 - 4*xc**3 - 2*xc + yc**4 + 4*yc**2 + 3
assert p.subs([(xc**i, yc**i) for i in [2, 4]]) == p_subbed
assert (x**4 + x**3 + x**2 + x + sqrt(x)).subs(x**2, y) == \
    sqrt(x) + x**3 + x + y**2 + y
assert x.subs(x**3, y) == x
assert (1/x).subs(x**2, 1) == 1/x
assert (1/(x**3)).subs(x**2, 1) == x**(-3)
assert (1/(x**3)).subs(x**4, 1) == x**(-3)
```
There is at least one test covering the nonnegative case.

The following diff passes all tests in core and gives a correct result for the OP integral:
```diff
diff --git a/sympy/core/power.py b/sympy/core/power.py
index 38c9987339..2c4a82b901 100644
--- a/sympy/core/power.py
+++ b/sympy/core/power.py
@@ -832,9 +832,9 @@ def _check(ct1, ct2, old):
                         as_int(pow, strict=False)
                         combines = True
                     except ValueError:
-                        combines = isinstance(Pow._eval_power(
-                            Pow(*old.as_base_exp(), evaluate=False),
-                            pow), (Pow, exp, Symbol))
+                        b, e = old.as_base_exp()
+                        # These conditions ensure that (b**e)**f == b**(e*f) for any f
+                        combines = b.is_positive and e.is_real or b.is_nonnegative and e.is_nonnegative
                     return combines, pow, None
                 else:
                     # With noncommutative symbols, substitute only integer powers
```

```python
In [1]: integrate((x**2 - Rational(1, 4))**2 * sqrt(1 - x**2), (x, -1, 1))
Out[1]: 
π 
──
32
```

**Comment by oscarbenjamin** [comment]
> The following diff passes all tests in core and gives a correct result for the OP integral:

It also fixes the integrals from #21671

**Comment by smichr** [comment]
> These lines make all real valued base expressions nonnegative:

b = abs(b) under the assumption that e is even. How specifically is this leading to a problem? This is applied, for example, for (x**2)**(3/2) when x is real to give `x**2*abs(x)`

**Comment by smichr** [comment]
> Question: Please read the attached file.

![image](https://user-images.githubusercontent.com/90703/132682611-01255280-e5ad-406e-a516-97827a1e253b.png)

The constant of integration can make either answer right, can't it? But in the context of integration don't want this because if we differentiate the answer we would like to get back the original expression. That won't happen for the expression involving `Abs`. The answer might be, then, to replace the `Abs(x)` with `x` as a clean-up step.

**Comment by jksuom** [comment]
> b = abs(b) under the assumption that e is even. How specifically is this leading to a problem? 

The problem was in connection with `subs`, but it looks like that will be fixed now.

**Comment by smichr** [comment]
> but it looks like that will be fixed now.

It is not clear to me why taking only a subset of the `_eval_pow` logic is the right thing to do in subs.

**Comment by MakotoTsukada** [comment]
If x in non-negative, x^2/3 is the right answer. If x is negative, since dx=-dy/(2sqrt(y)), the answer is -abs(x)^3/3=x^3/3.

**Comment by smichr** [comment]
The problem seems to be in how to handle the change of variables. If `y = x**2` then `x = +/-sqrt(y)`. I am not sure of the correct way to handle the two signs. Perhaps a change of variables should not be attempted when the inversion is not single-valued?

**Comment by oscarbenjamin** [comment]
I think the fix in #22052 basically just prevents the substitution.

There could be cases where it works but you would need the `+/-` to cancel out in some other way.

For a definite integral any substitution needs to have a uniquely defined inverse across the whole region e.g. with `y = x**2` the `1/sqrt(y)` is singular at `y=0`. That should immediately make it obvious that the integral needs to be split for that substitution to be used.

**Comment by smichr** [comment]
> any substitution needs to have a uniquely defined inverse across the whole region

So that is the root issue for integrate and #22052 is just helping `integrate` avoid the root issue as it arises with substitution, right?
